setuptools/dist.py

```diff
@@ -138,11 +139,15 @@ def invalid_unless_false(dist, attr, value):
     raise DistutilsSetupError(f"{attr} is invalid.")
 
 
-def check_requirements(dist, attr, value):
+@overload
+def check_requirements(dist, attr: str, value: set) -> NoReturn: ...
```

Comment:
I don't like this at all. We don't want to be declaring that `set` is valid, because this function is specifically designed to prevent `set`. I would normally say let's just drop the imperative check and let users rely on mypy to check their types, but I'm guessing it's unlikely that anyone is type-checking their `setup.py`. Also, I'm unsure if there are other code paths. For example, can a user supply a set using `pyproject.toml`? It seems not.

So maybe the best thing would be to drop this override and suppress mypy errors/warnings about its usage.

Reply:
> We don't want to be declaring that set is valid

That's what this overload is doing, *preventing* the use of `set` in `check_requirements` by indicating it'll never return (ie it'll always raise or exit the program).

There's some cases where it doesn't help (like if `check_requirement` is the last statement of a block), but without it `set` was already always valid anyway because it matches `_StrOrIter`.

One could also use the static `@deprecated` decorator on that overload to further re-enforce and message the user, but that's a bit cheesy (and wouldn't be clean as it'd require a runtime `typing_extension` import or alias)

Finally as per https://github.com/pypa/setuptools/pull/4575/files#r1733535055 , we can instead try to restrict the type of `value` further. For instance `Sequence[str], which already matches a type of nested strings incidentally, and would prevent passing `dict` or `set`.

![image](https://github.com/user-attachments/assets/6eb5d9b1-4cf8-46ea-99d9-f734623a5b9e)

Reply:
Thanks for the explanation. As long as IDEs aren't tempted to _suggest_ passing a `set` because of this annotation, I'm fine with it.

Reply:
> Thanks for the explanation. As long as IDEs aren't tempted to _suggest_ passing a `set` because of this annotation, I'm fine with it.

I just left home, but later I can take a look and show you what it looks like in VSCode/Pylance (especially when it comes to intellisense suggestions). As to not make assumptions here.

Reply:
Because of the necessary overload order, it still suggested sets first. Whilst it also showed it would never return `NoReturn` (return type), that didn't feel right. So I went with your other suggestion. I think that's good enough.